FEATURE: Edit preview mode support for Neos 9

mficzel opened the pull request:
The main change is that `edit` and `preview` action of the node controller are now separated with distinct constraints. This now allows for an mode to be edit and preview mode at the same time which is allowed by the configuration anyways.

- NodeController: has separate edit and preview actions that take the previewMode as argument
- NodeUriBuilder and LinkingService: will use preview / edit action once the main request is from the same action
- Neos.BackendHelper: has additional methods isEditMode, isPreviewMode and renderingModeCacheIdentifier

!!! The Neos.UI needs a change to provide the `editPreviewMode` parameter and call the edit or the preview action !!!

!!! This pr changes some behaviors slightly:

1. The distinction between edit and preview mode is now made based on the called action of the nodeController
2. The eel method `Neos.Node.inBackend(node)` is removed and is replaced with `Neos.Backend.isEditMode(request) || Neos.Backend.isPreviewMode(request)` this is non breaky as this method was never in an official release

Resolves: [#4086](https://github.com/neos/neos-development-collection/issues/4086)
Resolves partly: https://github.com/neos/neos-development-collection/issues/4396

**Upgrade instructions**

If fusion code relies on the edit preview mode beeing available via `node.context.currentRenderingMode` it has to be adjusted. Usually this means adjusting from code like `${node.context.currentRenderingMode.edit}` to `${Neos.Backend.isEditMode(request)}`.

**Review instructions**

Until the ui is adjusted this will always the preview as this was the existing method of the node controller. By setting the default value of the `editPreviewMode` on the `previewAction` to `rawContent` you can verify the rendering via different fusion pathes.

For now you can use the console to call the edit preview modes in your browser. You will have to replace the `node` with a node-address for your setup. You can then verify that the edit / preview mode is persisted during navigation after the initial call.

1. Open document in default preview mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/preview?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
2. Open document default edit mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
3. Open document in rawContent mode (edit)
`document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677&editPreviewMode=rawContent"` 

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

**Todos**
- [ ] required UI change https://github.com/neos/neos-ui/pull/3411
- [ ] `Neos.Backend.isEditMode` automated migration https://github.com/neos/rector/issues/21
- [ ] `Neos.Backend.isEditMode` migration in the UI https://github.com/neos/neos-ui/pull/3571

ahaeslich commented:
@mficzel This will probably create a merge conflict with neos/neos-development-collection#4074.

skurfuerst commented:
@mficzel @mhsdesign you let me know when we can review / merge this? :) all the best :)

mficzel commented:
@akurfuerst a little rebase aside this should be ready to go. Only problem is the ui change that is needed here https://github.com/neos/neos-ui/pull/3411

@ahaeslich do you remember why we did not add a description

mficzel commented:
@ahaeslich I added a description from memory. Did i miss something important.

I also merged the 9.0 branch into the state here. Looking OK so far and it seems to work fine for now.

mficzel commented:
Noticed the related PRs:

- https://github.com/neos/neos-ui/pull/3471
- https://github.com/neos/neos-development-collection/pull/4217

mficzel committed:
FEATURE: Implement edit preview mode for Neos 9

The main change is that edit and preview action are now separated with distinct constraints.

- NodeController: has separate edit and preview actions that take the previewMode as argument
- NodeUriBuilder and LinkingService: will use preview / edit action once the main request is from the same action
- Neos.BackendHelper: has additional methods isEditMode, isPreviewMode and editPreviewModeCacheIdentifier

mficzel committed:
TASK: Do not pass the edit preview mode to the rendering but. Instead the fusion path is set by the nodeController

The main difference is that this leads to edit preview modes now rendering shortcut nodes using the configured fusionPath

bwaidelich commented:
Any updates on this one?

mficzel committed:
TASK: Add case to rawContent fusion path to ensure shortcuts are rendered as before

This is necessary as the fusionPath is now set on the view directly if the edit preview mode defined one.

mficzel committed:
TASK: Make linter happy

mficzel commented:
@bwaidelich i consider this ready backend-wise. However the ui will not work until it is adjusted to this change as it will default to preview mode if unadjusted.

The really important changes are for now:
1. The edit and preview are now distinct actions of the NodeController now  -> this idea taken from the pr https://github.com/neos/neos-development-collection/pull/4217 from @daniellienert 
2. The current fusion path (if specified) is set on the view directly instead of switching in the rootCase.
3. The backend helper has new methods `isEditMode`, `isPreviewMode` and `renderingModeCacheIdentifier` that each take the `request` as argument.
4. The LinkingService and NodeUriBuilder will also check the current request to create links with in the current editPreviewMode.

If we agree on those changes in general we should adjust the UI as follows: https://github.com/neos/neos-ui/pull/3411

1. The ui has to links to current document via the edit or the preview action (since before there was only the preview action one currently ends up in preview mode)
2. The ui edit preview mode selector has to adjust to the possibility of modes beeing edit- and previewMode at the same time. (the Settings allowed that anyways)
3. Investigate why new content in `rawContent` mode leads to reloading the document in `inPlace` mode

For now you can use the console to call the edit preview modes in your browser. You will have to replace the node with a node-address for your setup. You can then verify that the edit / preview mode is persisted during navigation after the initial call.

1. Open document in default preview mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/preview?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677" `
2. Open document default edit mode `document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677"`
3. Open document in rawContent mode (edit)
`document.querySelector('iframe').src = "https://neos-development-distribution-9-0.ddev.site/neos/edit?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__995c9174-ddd6-4d5c-cfc0-1ffc82184677&editPreviewMode=rawContent"`

mficzel committed:
BUGFIX: Correct fallback to rawContentMode in backend to use new helpers

mhsdesign commented on Neos.Neos/Classes/Service/LinkingService.php:
```diff
@@ -377,7 +391,7 @@ public function createNodeUri(
             ->setFormat($format ?: $request->getFormat())
             ->setCreateAbsoluteUri($absolute)
-            ->uriFor($action, ['node' => $node], 'Frontend\Node', 'Neos.Neos');
+            ->uriFor($action, ['node' => $node], 'Frontend\Node', 'Neos.Neos') . '&grrr';
```
grrrr

mhsdesign commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+
+    public function renderingModeCacheIdentifier(ActionRequest $request): string
+    {
+        if (
+            $request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && ($request->getControllerActionName() === 'edit' || $request->getControllerActionName() === 'preview')
+        ) {
```
```suggestion
        if ($this->isEditMode($request) || $this->isPreviewMode($request)) {
```

mficzel replied:
This will check the controller and package twice ... not against it but also not a fan.

bwaidelich replied:
I'm also in favor of preventing domain logic duplication.
`self::isEditMode(...) || self::isPreviewMode(...)` is much easier to comprehend and those string comparisons come with no remarkable overhead

mhsdesign commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+            }
+            return $request->getControllerActionName();
+        } else {
+            return "show";
```
```suggestion
            return null;
```

instead?

mficzel replied:
:-/ that would mean to change the return type to ?string ... i personally like cache identifier as string.

bwaidelich replied:
I agree, nullable (return) types add complexity.
I'm just a bit unsure about the many magic strings in the code, but than again I think the actual value is not that important as long as it's unique, right?
nitpick: double quotes should be replaced though

mficzel replied:
Added single quotes as suggested.

mhsdesign commented on Neos.Neos/Configuration/Settings.yaml:
```diff
@@ -278,7 +278,7 @@ Neos:
           title: Live
         inPlace:
           isEditingMode: true
-          isPreviewMode: false
+          isPreviewMode: true
```
Your thesis:

defaultEditPreviewMode should have both to true

mficzel replied:
One could rephrase this as the ui has to switch to edit / preview depending on what the defaultMode is.

mficzel replied:
I agree this is not required anymore once the ui knows about the distinction of edit vs preview action. Currently this ensures that at least the preview mode is shown instead of an error.

mficzel committed:
TASK: Remove debug leftovers and ensure no content element wrapping is done in preview

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+    public function isEditMode(ActionRequest $request): bool
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
```
```suggestion
            && $request->getControllerName() === 'Frontend\\Node'
```

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+    public function isPreviewMode(ActionRequest $request): bool
+    {
+        return ($request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
```
```suggestion
            && $request->getControllerName() === 'Frontend\\Node'
```

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
         return $this->userService->getInterfaceLanguage();
     }
 
+    public function isEditMode(ActionRequest $request): bool
```
This method could be static as far as I can tell
```suggestion
    public static function isEditMode(ActionRequest $request): bool
```

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+        );
+    }
+
+    public function isPreviewMode(ActionRequest $request): bool
```
```suggestion
    public static function isPreviewMode(ActionRequest $request): bool
```

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
```diff
@@ -40,6 +41,39 @@ public function interfaceLanguage(): string
+        );
+    }
+
+    public function renderingModeCacheIdentifier(ActionRequest $request): string
```
```suggestion
    public static function renderingModeCacheIdentifier(ActionRequest $request): string
```

bwaidelich commented on Neos.Neos/Classes/Service/LinkingService.php:
```diff
@@ -367,7 +369,19 @@ public function createNodeUri(
         $uriBuilder->setRequest($request);
-        $action = $workspace && $workspace->isPublicWorkspace() && !$hiddenState->isHidden ? 'show' : 'preview';
+
+        if (
+            $request->getControllerPackageKey() === 'Neos.Neos'
+            && $request->getControllerName() === "Frontend\Node"
+            && in_array($request->getControllerActionName(), ['edit', 'preview'])
+        ) {
+            $action = $request->getControllerActionName();
+            if ($request->hasArgument('editPreviewMode')) {
+                $arguments['editPreviewMode'] = $request->getArgument('editPreviewMode');
+            }
+        } else {
```
We duplicate the logic here. Maybe something like this works:
```suggestion
        if (BackendHelper::isEditMode($request) || BackendHelper::isPreviewMode($request)) {
            $action = $request->getControllerActionName();
            if ($request->hasArgument('editPreviewMode')) {
                $arguments['editPreviewMode'] = $request->getArgument('editPreviewMode');
            }
        } else {
```

mficzel committed:
Apply suggestions from code review

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>
Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

mficzel commented:
I applied the suggested changes.

mhsdesign commented:
we should deprecate `Node.inBackend` as the concept from using the request is the correct one.

`node.workspace.isLive !== (!Node.inBackend)`

grebaldi commented:
This PR should cover the content collection, right?:
https://github.com/neos/neos-development-collection/blob/f4ed8369957209ba689cb2e1d9319e39716d6ac3/Neos.Neos/Resources/Private/Fusion/Prototypes/ContentCollection.fusion#L30

(see: neos/rector#23)

mficzel committed:
TASK: Remove the `Neos.Node.inBackend` and use `Neos.Backend.isEditMode()` && `Neos.Backend.isPreviewMode()`

`Neos.Node.inBackend` was intended to be a replacement for `node.context.inBackend` but the detection wether a node is rendered in the backend
is now done in the Backend helper by checking wich action has actually been called.

mficzel commented:
@grebaldi @mhsdesign good catch. Removed `Neos.Node.inBackend(node)` and replaced the usage with the `Neos.Backend.isEditMode(request) || Neos.Backend.isPreviewMode(request)`

grebaldi commented:
Nice one @mficzel, thx!

I'll patch this into the test distribution over at https://github.com/neos/neos-ui/pull/3569 and get back to you with a full review :)

mficzel commented:
@grebaldi zu needs this to get along with removed `Neos.Node.inBackend(node)` https://github.com/neos/neos-ui/pull/3571

mficzel commented:
@grebaldi any support with adjusting the ui-side to the changes in the backend node entry points would be very much apreciated. Pretty sure the fact that this can be hardly tested is pert of the reason why there are no reviews yet.

mhsdesign commented:
should `Neos.Backend.isPreviewMode` be true when you open the page externally?

![image](https://github.com/neos/neos-development-collection/assets/85400359/35efe3db-4921-4e1b-9079-b9dd7ada0ece)

(havent checked the behavior of this pr - just a reminder for me)

mficzel commented:
> should Neos.Backend.isPreviewMode be true when you open the page externally?

@mhsdesign Pretty sure yes ... that was basically what the pr https://github.com/neos/neos-development-collection/pull/4217 was about that first tried to introduce separate edit and preview actions. Since previously the user session decided wether an edit or a preview mode was active it was hard to be in edit mode but render a preview using the show preview link.

mhsdesign commented:
Im just wondering about caching.

When retrieving the pseudo `inBackend` information from the node (which actually only says if the node is not in the live workspace) caching works out of the box right?, because its a different node when `inBackend` changes.

But when taking the `request` we need to adjust the caching configuration so that this works?

```
renderer = ${Neos.Backend.isEditMode(request) ? 'edit' : (Neos.Backend.isPreviewMode(request) ? 'preview' : 'live')}
```

edit:

it seems you have done that in the global cache ids already:

```
renderingMode = ${Neos.Backend.renderingModeCacheIdentifier(request)}
```

grebaldi commented on Neos.Neos/Classes/Domain/Model/EditPreviewMode.php:
```diff
@@ -0,0 +1,43 @@
+        public readonly ?string $fusionPath,
+        public readonly bool $isEditMode,
+        public readonly bool $isPreviewMode
+    ) {
+    }
```
We were talking about this model. It should contain a check, so that `$isEditMode` and `$isPreviewMode` cannot both be true at once (and that one of them must be true).

mhsdesign replied:
dont say enum ;)

grebaldi replied:
I didn't :D

But yeah, that would probably be better.

grebaldi commented on Neos.Neos/Classes/Domain/Repository/EditPreviewModeRepository.php:
```diff
@@ -0,0 +1,44 @@
+        if (array_key_exists($name, $this->editPreviewModeConfigurations)) {
+            return EditPreviewMode::fromNameAndConfiguration($name, $this->editPreviewModeConfigurations[$name]);
+        }
+        throw new InvalidEditPreviewModeException(sprintf('"%s" is not a valid editPreviewMode', $name), 1683790077);
```
```suggestion
        throw new InvalidEditPreviewModeException(sprintf('An editPreviewMode with the name "%s" does not exist', $name), 1683790077);
```

grebaldi commented on Neos.Neos/Classes/Fusion/Helper/BackendHelper.php:
I'm very unsure whether it's wise to check for the current edit/preview mode via this EEL-Helper and by analyzing the request. As a consequence of this strategy, this is what needs to be done for out-of-band rendering in the UI:

https://github.com/neos/neos-ui/blob/0a858bb1cd5a31e4143f9b06144b0704c3acaad7/Classes/Domain/Model/Feedback/Operations/ReloadContentOutOfBand.php#L136-L146
*(These lines show the creation of a fake controller context, so this EEL-Helper is convinced to be executed in a different request than it actually is)*

Intuitively, I'd say it'd be better to just throw the current `EditPreviewMode` object into the global fusion context, so it can just be used like

```fusion
someKey.@if.inEditMode = ${editPreviewMode.isEditMode}
```

On the other hand, that fake controller context may not be the worst idea after all, because usually integrators wouldn't expect (and shouldn't notice) their code to go through such different execution paths.

However, semantically it seems very arbitrary to me to bind the current edit preview mode to a specific controller action...

mhsdesign replied:
Yes i rather have this additional context too. I remember that we discussed this at the Dresden sprint, but the discussion came up that in fusion we can't currently add new context variables because users would have to account for them in caching.

There is a way though: We need to pass the context variable behind fusions back like `request` so its always available in eel.

I will prepare a draft pr to show you what i mean ;)

mhsdesign replied:
@grebaldi i prepared a concept over here: https://github.com/neos/neos-development-collection/pull/4425

this way we open the api of the runtime to allow other variables like `request` who are always present (in the instance of the runtime) and thus not affected by caching.

This would allow us to add a global `editPreviewMode` object with no dependency on the request. ;)

mhsdesign replied:
```php
new Runtime(
  defaultContextVariables: [
      'request' => $request,
      'Neos' => [
        'UiMode' => new UiModeHelper($isBackend)
      ]
  ]
)
```

grebaldi commented on Neos.Neos/Classes/Service/LinkingService.php:
```diff
@@ -367,7 +370,15 @@ public function createNodeUri(
         $uriBuilder->setRequest($request);
-        $action = $workspace && $workspace->isPublicWorkspace() && !$hiddenState->isHidden ? 'show' : 'preview';
+
+        if (BackendHelper::isEditMode($request) || BackendHelper::isPreviewMode($request)) {
```
I'd prefer to move this closer to the domain. Looks like a job for the `EditPreviewModeRepository`:

```php
final class EditPreviewModeRepository
{
    public function findOneByActionRequest(ActionRequest $request): ?EditPreviewMode;
}
```

So that this check translates to 

```suggestion
        $editPreviewMode = $this->editPreviewModeRepository->findOneByActionRequest($request);
        if ($editPreviewMode) {
```

Depending on how we move forward with the `BackendHelper` itself, it should of course use the same method to retrieve the proper `EditPreviewMode` from the request.

grebaldi commented on Neos.Neos/Classes/Controller/Frontend/NodeController.php:
```diff
@@ -119,7 +129,38 @@ class NodeController extends ActionController
-    public function previewAction(string $node): void
+    public function previewAction(string $node, string $editPreviewMode = null): void
+    {
+        $editPreviewModeObject = $editPreviewMode ? $this->editPreviewModeRepository->findByName($editPreviewMode) : $this->editPreviewModeRepository->findDefault();
+        if ($editPreviewModeObject->isPreviewMode === false) {
+            throw new InvalidEditPreviewModeException(sprintf('"%s" is not a preview mode', $editPreviewMode), 1683127314);
```
I would also prefer to move this logic closer to the domain, by letting the `EditPreviewModeRepository` provide more specific retrieval methods:

```php
final class EditPreviewModeRepository
{
    /** @throws NotAnEditModeException */
    public function findOneEditModeByName(string $name): ?EditPreviewMode;

    /** @throws DefaultEditModeNotFoundException */
    public function findDefaultEditMode(): EditPreviewMode;

    /** @throws NotAPreviewModeException */
    public function findOnePreviewModeByName(string $name): ?EditPreviewMode;

    /** @throws DefaultPreviewModeNotFoundException */
    public function findDefaultPreviewMode(): EditPreviewMode;
}
```

grebaldi commented on Neos.Neos/Classes/Controller/Frontend/NodeController.php:
```diff
@@ -119,7 +129,38 @@ class NodeController extends ActionController
+     */
+    public function editAction(string $node, ?string $editPreviewMode = null): void
+    {
+        $editPreviewModeObject = $editPreviewMode ? $this->editPreviewModeRepository->findByName($editPreviewMode) : $this->editPreviewModeRepository->findDefault();
+        if ($editPreviewModeObject->isEditMode === false) {
+            throw new InvalidEditPreviewModeException(sprintf('"%s" is not an edit mode', $editPreviewModeObject->name), 1683127295);
```
Likewise

grebaldi reviewed (commented):
Hi @mficzel,

thanks for taking care of this :) I was able to successfully test this change in the scope of https://github.com/neos/neos-ui/pull/3569. 

Therein I stumbled upon some conceptual questions that we've already talked about briefly. The major questions are imho:

- Should it be possible for an `EditPreviewMode` to be both `edit`- AND `preview`-mode
  - Within this PR this is currently the default for `inPlace`
  - As we've spoken about this, my impression was that we generally agree that this shouldn't be possible, but that we need a solution for the UI to behave correctly, because of the change of endpoints
- Should the current `EditPreviewMode` be really dependent on the current action request?
  - The alternative would be to put the current `EditPreviewMode` object into the global fusion context
  - There are pros and cons for both solutions

All other comments I left are just some opinionated suggestions :)

So, in conclusion, we need to clear up those questions and then this should be good to go imho. But we need to coordinate closely, because once this is merged, the UI will cease to work, unless https://github.com/neos/neos-ui/pull/3569 is merged as well.

mficzel commented:
Am AFK for the next three weeks. Feel free to add all the changes that are necessary to this pr.

mficzel commented:
Close in favor of https://github.com/neos/neos-development-collection/pull/4505 please let the branch live for a bit in case it is needed as comparison.
